api/client/samples/similar_regions_Frechet/sr.py

```diff
@@ -0,0 +1,560 @@
+import os, glob
+import tempfile
+import pickle
+import logging
+from datetime import date
+
+import numpy as np
+from scipy.linalg import sqrtm
+import pandas as pd
+from tqdm import tqdm
+
+from sklearn.neighbors import BallTree, DistanceMetric
+from api.client.batch_client import BatchClient, BatchError
+from api.client.lib import get_default_logger
+#from api.client.samples.similar_regions.similar_region_state import SimilarRegionState
+from sklearn.metrics.pairwise import euclidean_distances
+
+""" API Config """
+API_HOST = 'api.gro-intelligence.com'
+ACCESS_TOKEN = os.environ['GROAPI_TOKEN']
+
+""" Cache dir and file names """
+# cache directory should already exist when SR object is created (otherwise /tmp will be used)
+CACHE_PATH = "sr_cache"
+REGION_INFO_CACHE = "region_info"
+DATA_CACHE = "region_data_"
+PICKLE_PROTOCOL = 4
+REGIONS_PER_QUERY = 100
+MAX_RETRY_FAILED_REGIONS = 3
+OK_TO_PROCEED_REGION_FRACTION = 0.1 # we want at least 10% of all desired region present in the final data matrix
+
+class SimilarRegion(object):
+
+    def __init__(self, metric_properties,
+                 search_region=0,
+                 update_mode="no",
```

Comment:
i think we have too many options here.   Lets' just make it load a pickle file with the whole object and recover a prior state.  so basically replace update_mode, and region_info_reload and data_dir with just a file path.   If the file is not given  it just starts empty. if the file is given but doesn't exist it should  fail.

Reply:
done in 73d773ddab14619279ab7152c90922d827439949

i've simplified it so it always uses the cache if present and updates if anything needed is missing, both region info and data. the user also can just delete the cache or point to a new  dir to force reload, which is also convenient when working on different scenarios

Reply:
Maybe I am missing something, but this sounds like pickling a (previously created) SR object. But we don't need this functionality within SR class - just do pickle.dump/pickle.load as with any other class. And some sort of "partial pickling" is probably an unnecessary complication.

 I agree that  there are quite a few parameters here, but this is a complex object and at some point it needs to be generated from scratch, so even if we pickle most of the object somehow, we still need to specify full configuration from time to time. However, for a "casual" user defaults should be fine. In fact, maybe we can remove region_info_reload and drop_mode altogether (always use their default values).

 I'd prefer to keep update_mode - "starting empty" means full reload and, given the time it takes, this should be requested explicitly (via update_mode="reload").  data_dir is mostly to have the ability to work with multiple data caches. This is convenient sometimes, but might not be needed by a typical user, so can be dropped as well (i.e. we will always use ./sr_cache directory). We can go even a step further and kill t_int_per_year option (always use 52) since users probably not going to play with this (and it required data re-download). 
 
With all these together we will have minimal form as something like
    def __init__(self, metric_properties,
                 search_region=0,
                 update_mode="no",
                 metric_instance=None)
In fact, with only this few parameters, data_dir could probably be kept as well as it does come convenient sometimes.

Overall proposal is therefore to remove only region_info_reload, t_int_per_year, and drop_mode. I think this should be sufficient. Note that these options can still be accessed via globals in the code. I guess we expect advanced users to be able to modify code if needed (since we are releasing the code in the first place)

Reply:
My earlier comment was a bit to succinct... i did keep data_dir and the separate region and date cache files as before.  Here's what we have now
- always loads saved region info  if it finds it in the specified data_dir
- always updates as necessary if region is missing
- load data if any is missing by default  (but still supports the other "drop_modes" since it's already done)

Also as a bonus, I've made it so that

-  supports comparison across different region trees.. E.g. you can pick a district in US and find similar provinces in Europe.
-  when the comparison regions changed, it is automatically updated. so basically the user only needs to call similar_to() no matter what the regions are and never worry about re-initializing.